### PR TITLE
Track indexed partitions to avoid unnecessary reindexing

### PR DESCRIPTION
**Pull request by taburet**
EE part: https://github.com/hazelcast/hazelcast-enterprise/pull/2543

This change fixes the following issues:

1. When a user invokes IMap.addIndex, AddIndexOperation is sent to all
members and the index in question is fully rebuilt even if it's already
populated. Typically, such addIndex invocations are performed from user
code for each member joining a cluster during the member initialization.
If the cluster is storing a lot of data, rebuilding all of the indexes
may be a very costly operation.

2. When a new member joins a cluster, a local map proxy object is
created on that new member. During the proxy initialization,
AddIndexOperation is sent to all cluster members for each index
configured by the new member XML configuration to ensure the indexes
are present on all members (see MapProxySupport.initializeIndexes).
Basically, we are rebuilding all of the indexes on all of the members.

This change tracks a set of partitions indexed by each index. If a
partition is already indexed by a certain index, the reindexing is
skipped.

Fixes: https://github.com/hazelcast/hazelcast/issues/13964

### CODE REVIEW

**Comment by taburet** [comment]
run-lab-run

**Comment by taburet** [comment]
run-lab-run

File: hazelcast/src/main/java/com/hazelcast/query/impl/IndexImpl.java [hunk 1]
```diff
@@ -16,247 +16,51 @@
...
-    private final boolean ordered;
-    private final IndexCopyBehavior copyBehavior;
-    private final PerIndexStats stats;
+/**
+ * Provides implementation of on-heap indexes.
+ */
+public class IndexImpl extends AbstractIndex {
 
-    private volatile TypeConverter converter;
+    private final Set<Integer> indexedPartitions = Collections.newSetFromMap(new ConcurrentHashMap<Integer, Boolean>());
```
Comment (ahmetmircik) [inline hazelcast/src/main/java/com/hazelcast/query/impl/IndexImpl.java]: can be a fixed size AtomicIntegerArray or something array based like that?

**Comment by ahmetmircik** [comment]
Can there be still re-indexing if i define indexes on xml declaratively, and later call addIndex for the same attribute? Do we handle this?

File: hazelcast/src/main/java/com/hazelcast/query/impl/IndexImpl.java (hunk 1 above)
Comment (taburet) [reply hazelcast/src/main/java/com/hazelcast/query/impl/IndexImpl.java]: Yes, it can, but there is no that much value in doing so: (1) we would waste `partition_count` - `partition_count` / `member_count` slots in the array and (2) execution paths (index addition and migration) touching the set are not performance critical.

Another approach is creating something like `Concurrent/AtomicBitSet`, so we may pack 271 partitions in 9 ints or 5 longs, but I don't think this task deserves such complexity :)

Comment (ahmetmircik) [reply hazelcast/src/main/java/com/hazelcast/query/impl/IndexImpl.java]: ok makes sense
